completions/make.lua

```diff
@@ -0,0 +1,145 @@
+local special_targets = {
+    ['.PHONY'] = true,
+    ['.SUFFIXES'] = true,
+    ['.DEFAULT'] = true,
+    ['.PRECIOUS'] = true,
+    ['.INTERMEDIATE'] = true,
+    ['.SECONDARY'] = true,
+    ['.SECONDEXPANSION'] = true,
+    ['.DELETE_ON_ERROR'] = true,
+    ['.IGNORE'] = true,
+    ['.LOW_RESOLUTION_TIME'] = true,
+    ['.SILENT'] = true,
+    ['.EXPORT_ALL_VARIABLES'] = true,
+    ['.NOTPARALLEL'] = true,
+    ['.ONESHELL'] = true,
+    ['.POSIX'] = true,
+    ['.NOEXPORT'] = true,
+    ['.MAKE'] = true,
+}
+
+local function extract_target(line, last_line, targets, allow_pathlike)
+    -- Ignore comment lines.
+    if line:find('#', 1, true) then
+        return
+    end
+
+    -- Ignore when not a target (is this only for GNU make?).
+    if last_line:find('# Not a target') then
+        return
+    end
+
+    -- Extract possible target.
+    local p = (
+        line:match('^([^%s]+):$') or -- When target has no deps.
+        line:match('^([^%s]+): '))   -- When target has deps.
+    if not p then
+        return
+    end
+
+    -- Ignore targets with pattern
+    if p:find('%%') then
+        return
+    end
+
+    -- Ignore special targets.
+    if special_targets[p] then
+        return
+    end
+
+    -- Maybe ignore path-like targets.
+    local mt
+    if allow_pathlike then
+        if not p:find('[/\\]') then
+            mt = 'alias'
+        end
+    else
+        if p:find('[/\\]') then
+            return
+        end
+    end
+
+    -- Add target.
+    table.insert(targets, { match = p, type = mt })
+end
+
+-- Sort comparator to sort pathlike targets last.
+local function comp_target_sort(a, b)
+    local a_alias = (a.type == 'alias')
+    local b_alias = (b.type == 'alias')
+    if a_alias ~= b_alias then
+        return a_alias
+    else
+        return string.comparematches(a.match, b.match)
+    end
+end
+
+local function get_targets(word, word_index, line_state, builder, user_data) -- luacheck: no unused
+    local make_cmd = '"' .. line_state:getword(line_state:getcommandwordindex()) .. '" -p -q -r'
+    if user_data and user_data.makefile then
+        make_cmd = make_cmd .. ' -f "' .. user_data.makefile .. '"'
+    end
+
+    local file = io.popen('2>nul ' .. make_cmd)
+    if not file then
+        return
+    end
+
+    local targets = {}
+    local last_line = ''
+
+    local allow_pathlike = os.getenv('INCLUDE_PATHLIKE_MAKEFILE_TARGETS') and true
+
+    -- Extract targets to be included.
+    for line in file:lines() do
+        extract_target(line, last_line, targets, allow_pathlike)
+        last_line = line
+    end
+
+    file:close()
+
+    -- If pathlike targets are allowed to be included, sort them last.
+    if allow_pathlike and string.comparematches then
+        table.sort(targets, comp_target_sort)
+        if builder.setnosort then
+            builder:setnosort()
+        end
+    end
+
+
+    return targets
+end
+
+local function onarg_flags(arg_index, word, word_index, line_state, user_data)
```

Comment:
This function has been moved.  Can you keep the code in the same relative order as in the nmake.lua file, so that comparing the files has fewer unnecessary diff chunks?